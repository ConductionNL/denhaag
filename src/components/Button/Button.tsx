import { Button as MaterialButton } from "@material-ui/core";
import React, {ReactNode} from "react";

export default function () {
    return <MaterialButton/>;
}

export type ButtonProps = {
  /**
   * Simple click handler
   */
  onClick?: () => void;

  /**
   * Color for the component
   */
  color: "primary" | "secondary" | "default" | undefined;

  /** 
   * Button variant
   */
  variant: "outlined" | "contained" | "text" | undefined;
  children: ReactNode;
};

/**
 * Primary UI component for user interaction
 */ 

export const Button = ({
  color,
  variant,
  onClick,
  children,
  ...props
}: ButtonProps) => {
  return (
    <MaterialButton onClick={onClick} variant={variant} color={color} {...props}>
      {children}
    </MaterialButton>
  )
}
