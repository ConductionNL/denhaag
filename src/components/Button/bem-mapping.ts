export const classes = {
  root: 'denhaag-button',
  label: 'denhaag-button__label',
  contained: 'denhaag-button--primary-action',
  outlined: 'denhaag-button--secondary-action',
  disabled: 'denhaag-button--disabled',
  sizeMedium: 'denhaag-button--medium',
  sizeLarge: 'denhaag-button--large'
}
