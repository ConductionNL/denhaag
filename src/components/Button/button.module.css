.denhaag-button {
  --denhaag-button-padding-inline: 16px;
  --denhaag-button-border-radius: 3px;
  --denhaag-button-padding-block: 8px;
  --denhaag-button-focus-border-style: dashed;
  --denhaag-button-focus-border-color: var(--denhaag-ocher-4);
  --denhaag-button-focus-border-width: 0.125rem;
  --denhaag-button-font-family: sans-serif;

  --denhaag-button-primary-action-background-color: var(--denhaag-green-3);
  --denhaag-button-primary-action-color: var(--denhaag-white);
  --denhaag-button-primary-action-hover-background-color: var(--denhaag-green-4);
  --denhaag-button-primary-action-hover-color: var(--denhaag-white);
  --denhaag-button-primary-action-disabled-background-color: var(--denhaag-grey-2);
  --denhaag-button-primary-action-disabled-color: var(--denhaag-white);

  --denhaag-button-secondary-action-background-color: var(--denhaag-white);
  --denhaag-button-secondary-action-color: var(--denhaag-green-3);
  --denhaag-button-secondary-action-border-color: var(--denhaag-green-3);
  --denhaag-button-secondary-action-border-width: 0.0625rem;
  --denhaag-button-secondary-action-hover-background-color: var(--denhaag-white);
  --denhaag-button-secondary-action-hover-color: var(--denhaag-green-4);
  --denhaag-button-secondary-action-hover-border-color: var(--denhaag-green-4);
  --denhaag-button-secondary-action-disabled-background-color: var(--denhaag-white);
  --denhaag-button-secondary-action-disabled-color: var(--denhaag-grey-2);
  --denhaag-button-secondary-action-disabled-border-color: var(--denhaag-grey-2);

  --denhaag-button-medium-size-padding-block: var(--denhaag-button-padding-block);
  --denhaag-button-medium-size-padding-inline: var(--denhaag-button-padding-inline);
  --denhaag-button-large-size-padding-block: 10px;
  --denhaag-button-large-size-padding-inline: 20px;

  font-family: var(--denhaag-button-font-family);
  padding-inline-start: var(--denhaag-button-padding-inline);
  padding-inline-end: var(--denhaag-button-padding-inline);
  padding-block-start: var(--denhaag-button-padding-block);
  padding-block-end: var(--denhaag-button-padding-block);
  border-radius: var(--denhaag-button-border-radius);
  box-sizing: border-box;
}

.denhaag-button--primary-action {
  background-color: var(--denhaag-button-primary-action-background-color);
  color: var(--denhaag-button-primary-action-color);
}

.denhaag-button--primary-action.denhaag-button--hover, .denhaag-button--primary-action:hover {
  background-color: var(--denhaag-button-primary-action-hover-background-color);
  color: var(--denhaag-button-primary-action-hover-color);
}

.denhaag-button--primary-action.denhaag-button--disabled {
  background-color: var(--denhaag-button-primary-action-disabled-background-color);
  color: var(--denhaag-button-primary-action-disabled-color);
}

.denhaag-button--secondary-action {
  background-color: var(--denhaag-button-secondary-action-background-color);
  color: var(--denhaag-button-secondary-action-color);
  border-width: var(--denhaag-button-secondary-action-border-width);
  border-style: solid;
  border-color: var(--denhaag-button-secondary-action-border-color);
}

.denhaag-button--secondary-action.denhaag-button--hover, .denhaag-button--secondary-action:hover {
  background-color: var(--denhaag-button-secondary-action-hover-background-color);
  color: var(--denhaag-button-secondary-action-hover-color);
  border-color: var(--denhaag-button-secondary-action-hover-border-color);
}

.denhaag-button--secondary-action.denhaag-button--disabled {
  background-color: var(--denhaag-button-secondary-action-disabled-background-color);
  color: var(--denhaag-button-secondary-action-disabled-color);
  border-color: var(--denhaag-button-secondary-action-disabled-border-color);
}

.denhaag-button--focus, .denhaag-button:focus {
  outline: var(--denhaag-button-focus-border-color) var(--denhaag-button-focus-border-style) var(--denhaag-button-focus-border-width);
}

.denhaag-button--medium {
  padding-inline: var(--denhaag-button-medium-size-padding-inline);
  padding-block: var(--denhaag-button-medium-size-padding-block);
  font-size: var(--denhaag-fontsize-base);

}

.denhaag-button--large {
  padding-inline: var(--denhaag-button-large-size-padding-inline);
  padding-block: var(--denhaag-button-large-size-padding-block);
  font-size: var(--denhaag-fontsize-lg);
}

.denhaag-button--medium.denhaag-button--secondary-action {
  padding-inline: calc(var(--denhaag-button-padding-inline) - var(--denhaag-button-secondary-action-border-width));
  padding-block: calc(var(--denhaag-button-padding-block) - var(--denhaag-button-secondary-action-border-width));
}

.denhaag-button--large.denhaag-button--secondary-action {
  padding-inline: calc(var(--denhaag-button-large-size-padding-inline) - var(--denhaag-button-secondary-action-border-width));
  padding-block: calc(var(--denhaag-button-large-size-padding-block) - var(--denhaag-button-secondary-action-border-width));
}
