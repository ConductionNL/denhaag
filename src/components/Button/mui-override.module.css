.denhaag-button {
  box-shadow: none;
  transition: none;
}

.denhaag-button--hover, .denhaag-button:hover {
  box-shadow: none;
}

.denhaag-button--focus, .denhaag-button:focus {
  box-shadow: none;
}
