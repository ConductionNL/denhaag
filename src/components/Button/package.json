{
  "name": "@gemeente-denhaag/button",
  "description": "A Button component",
  "author": "Municipality of The Hague",
  "license": "EUPL-1.2",
  "main": "Button.tsx",
  "module": "Button.tsx",
  "version": "0.1.7",
  "repository": {
    "type": "git",
    "url": "https://github.com/Gemeente-DenHaag/denhaag-component-library.git",
    "directory": "src/components/Button"
  },
  "bugs": "https://github.com/Gemeente-DenHaag/denhaag-component-library/issues",
  "dependencies": {
    "@gemeente-denhaag/baseprops": "^0.1.6",
    "@gemeente-denhaag/basestyles": "^0.1.0",
    "@material-ui/core": "^4.11.0"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f",
  "exports": "./dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -b",
    "clean": "yarn rimraf dist tsconfig.tsbuildinfo"
  },
  "peerDependencies": {
    "react": "^17.0.1"
  }
}
