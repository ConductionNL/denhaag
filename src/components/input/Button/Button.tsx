import MaterialButton from '@material-ui/core/Button'
import React from 'react'

export interface ButtonProps {
  /**
   * Simple click handler
   */
  onClick?: () => void

  /**
   * Color for the component
   */
  color?: 'primary' | 'secondary' | 'default'

  /**
   * The url to link to when the button is clicked.
   * If defined, an a element will be used as the root node.
   */
  href?: string

  /**
   * Size of the component
   */
  size?: 'small' | 'medium' | 'large'

  /**
   * Disables Button
   */
  disabled?: boolean

  /**
   * Button variant
   */
  variant?: 'outlined' | 'contained' | 'text'

  /**
   * Content of the component.
   */
  children?: React.ReactNode
}

/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  return <MaterialButton {...props}>{props.children}</MaterialButton>
}

/**
 * Default export for Button
 */
export default Button
