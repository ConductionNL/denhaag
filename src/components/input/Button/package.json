{
  "name": "@gemeente-denhaag/button",
  "version": "0.1.1",
  "description": "A Button component",
  "author": "Municipality of The Hague",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/Gemeente-DenHaag/nlds-react-components.git",
    "directory": "src/components/input/Button"
  },
  "bugs": "https://github.com/Gemeente-DenHaag/nlds-react-components/issues",
  "dependencies": {
    "@gemeente-denhaag/baseprops": "^0.1.1",
    "@material-ui/core": "^4.11.0"
  }
}
