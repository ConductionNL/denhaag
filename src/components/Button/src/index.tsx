import MaterialButton, { ButtonTypeMap } from '@material-ui/core/Button';
import React from 'react';
import BaseProps from '@gemeente-denhaag/baseprops';
import { StylesProvider } from '@material-ui/core';
import { classes } from './bem-mapping';
import '@gemeente-denhaag/basestyles';
import './mui-override.module.css';
import './button.module.css';

export interface ButtonProps extends BaseProps {
  /**
   * Simple click handler
   */
  onClick?: () => void;
  /**
   * The url to link to when the button is clicked.
   * If defined, an a element will be used as the root node.
   */
  href?: string;

  /**
   * Size of the component
   */
  size?: 'medium' | 'large';

  /**
   * Disables Button
   */
  disabled?: boolean;

  /**
   * Button variant
   */
  variant?: 'primary-action' | 'secondary-action';

  /**
   * HTML type prop
   */
  type?: 'button' | 'submit' | 'reset';

  /**
   * Icon placed at the start of the button
   */
  startIcon?: React.ReactNode;

  /**
   * Icon placed at the end of the button
   */
  endIcon?: React.ReactNode;
}

/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  const sizeClass = `denhaag-button--${props.size ?? 'medium'}`;
  let muiVariant: ButtonTypeMap['props']['variant'] = 'contained';

  switch (props.variant) {
    case 'primary-action':
      muiVariant = 'contained';
      break;
    case 'secondary-action':
      muiVariant = 'outlined';
      break;
  }

  return (
    <StylesProvider injectFirst>
      <MaterialButton {...props} classes={classes} className={sizeClass} variant={muiVariant} disableRipple>
        {props.children}
      </MaterialButton>
    </StylesProvider>
  );
};

/**
 * Default export for Button
 */
export default Button;
