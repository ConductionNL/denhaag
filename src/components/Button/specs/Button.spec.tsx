import * as React from 'react';
import { mount } from '@cypress/react';
import Button, { ButtonProps } from '../src/index';

function terminalLog(violations) {
  cy.task(
    'log',
    `${violations.length} accessibility violation${violations.length === 1 ? '' : 's'} ${
      violations.length === 1 ? 'was' : 'were'
    } detected`,
  );

  const violationData = violations.map(({ id, impact, description, nodes }) => ({
    id,
    impact,
    description,
    nodes: nodes.length,
  }));

  cy.task('table', violationData);
}

describe('Button tests', () => {
  it('can be clicked', () => {
    const props: ButtonProps = {
      onClick: () => {
        console.log('Clicked');
      },
    };

    const spy = cy.spy(props, 'onClick').as('Button click');

    mount(<Button onClick={props.onClick}>Test button</Button>);

    cy.get('button')
      .contains('Test button')
      .click()
      .then(() => expect(spy).to.have.been.calledOnce);
  });

  it('cannot be clicked when disabled', () => {
    const props: ButtonProps = {
      onClick: () => {
        console.log('Clicked');
      },
    };

    const spy = cy.spy(props, 'onClick').as('Button click');

    mount(
      <Button onClick={props.onClick} disabled>
        Test button
      </Button>,
    );

    cy.get('button')
      .click({ force: true })
      .then(() => expect(spy).to.not.have.been.called);
  });

  it('does not violate any accessibility rules', () => {
    mount(<Button>Test button</Button>);
    cy.injectAxe();
    cy.checkA11y('button', null, terminalLog);
    cy.get('button').focus();
    cy.checkA11y('button', null, terminalLog);
    cy.get('button').click();
    cy.checkA11y('button', null, terminalLog);
  });

  it('matches snapshots', () => {
    mount(
      <div id="wrapper">
        <Button>Test button</Button>
        <Button variant="primary-action">Primary button</Button>
        <Button variant="secondary-action">Secondary button</Button>
        <Button
          startIcon={
            <svg width="24" height="24" viewBox="0 0 24 24">
              <path d="M0 0 24 24Z" fill="currentColor"></path>
            </svg>
          }
        >
          Icon Left Button
        </Button>
        <Button
          variant="secondary-action"
          endIcon={
            <svg width="24" height="24" viewBox="0 0 24 24">
              <path d="M0 0 24 24Z" fill="currentColor"></path>
            </svg>
          }
        >
          Icon Right Button
        </Button>
        <Button size="medium">Medium button</Button>
        <Button size="large">Large button</Button>
        <Button disabled>Disabled button</Button>
      </div>,
    );

    cy.get('#wrapper').toMatchSnapshot();
  });
});
