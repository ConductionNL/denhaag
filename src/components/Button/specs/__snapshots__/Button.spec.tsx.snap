exports[`Button tests > matches snapshots #0`] = `
<div id="wrapper">
  <button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--medium"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      >Test button</span
    ></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--medium"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      >Primary button</span
    ></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-outlined denhaag-button--secondary-action denhaag-button--medium"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      >Secondary button</span
    ></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--medium"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      ><span
        class="MuiButton-startIcon denhaag-button__start-icon MuiButton-iconSizeMedium denhaag-button__icon-size-medium"
        ><svg width="24" height="24" viewBox="0 0 24 24">
          <path d="M0 0 24 24Z" fill="currentColor"></path></svg></span
      >Icon Left Button</span
    ></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-outlined denhaag-button--secondary-action denhaag-button--medium"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      >Icon Right Button<span
        class="MuiButton-endIcon denhaag-button__end-icon MuiButton-iconSizeMedium denhaag-button__icon-size-medium"
        ><svg width="24" height="24" viewBox="0 0 24 24">
          <path d="M0 0 24 24Z" fill="currentColor"></path></svg></span
    ></span></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--medium"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      >Medium button</span
    ></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--large"
    tabindex="0"
    type="button"
  >
    <span class="MuiButton-label denhaag-button__label"
      >Large button</span
    ></button
  ><button
    class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--medium Mui-disabled denhaag-button--disabled Mui-disabled"
    tabindex="-1"
    type="button"
    disabled=""
  >
    <span class="MuiButton-label denhaag-button__label">Disabled button</span>
  </button>
</div>
`;
